Rstudio Weak Credentials Plugin

leonardo-doyensec opened the pull request:
Created a plugin for testing weak credentials in Rstudio Server

dawidg-doyen commented on google/detectors/credentials/generic_weak_credential_detector/src/main/java/com/google/tsunami/plugins/detectors/credentials/genericweakcredentialdetector/testers/rstudio/RStudioCredentialTester.java:
```diff
@@ -0,0 +1,259 @@
+    }
+  }
+
+  // This function is the same used from the javascript code in order to base64 encode the hexadecimal ciphertext. 
```
This comment line is too long. Must be up to 100 char. Rephrase to something shorter like: 

```// This function base64 encodes provided cipertext string in hex```

dawidg-doyen commented on google/detectors/credentials/generic_weak_credential_detector/src/main/java/com/google/tsunami/plugins/detectors/credentials/genericweakcredentialdetector/testers/rstudio/RStudioCredentialTester.java:
```diff
@@ -0,0 +1,259 @@
+          InvalidKeySpecException,
+          IOException,
+          NoSuchProviderException {
+    // encrypting with RSA PCKS#1 version 2
```
Start comments with a capital letter and the same tense as the others ("Encrypt data with...")

dawidg-doyen commented on google/detectors/credentials/generic_weak_credential_detector/src/main/java/com/google/tsunami/plugins/detectors/credentials/genericweakcredentialdetector/testers/rstudio/RStudioCredentialTester.java:
```diff
@@ -0,0 +1,259 @@
+          url, credential.username(), credential.password().orElse(""));
+      response = sendRequestWithCredentials(url, credential, exponent, modulus);
+
+      if (response.headers().get("Set-Cookie").isPresent()
```
startsWith could cause false negatives here I think. Perhaps there's another cookie with a different name (e.g. CSRF token) in front of the `user-id`.  Probably safer to use `.contains`

dawidg-doyen commented on google/detectors/credentials/generic_weak_credential_detector/src/main/java/com/google/tsunami/plugins/detectors/credentials/genericweakcredentialdetector/testers/rstudio/RStudioCredentialTester.java:
```diff
@@ -0,0 +1,259 @@
+    sb.append(credential.password().get());
+    byte[] cipherData = cipher.doFinal(sb.toString().getBytes());
+
+    // converting the ciphertext to hex
```
Fix tense/capital letter as above

dawidg-doyen commented on google/detectors/credentials/generic_weak_credential_detector/src/main/java/com/google/tsunami/plugins/detectors/credentials/genericweakcredentialdetector/testers/rstudio/RStudioCredentialTester.java:
```diff
@@ -0,0 +1,259 @@
+  }
+
+  // This function is the same used from the javascript code in order to base64 encode the hexadecimal ciphertext. 
+  private String hexToBase64(String hex) {
```
I think base encoding is already implemented in `com.google.common.io.BaseEncoding`. Have a look at `/tsunami-security-scanner-plugins/community/detectors/bigip_cve_2022_1388`. You should be able to remove it and replace with `BaseEncoding.base64().encode` to make the code shorter.

leonardo-doyensec replied:
I've tried to replace it with the function that you have said, but it's not working. Apparently rstudio performs some conversion from hex before converting it to base64. The function used in the code, is the translation of the one used in the client side code by rstudio. Let me know if i need to find an alternative

dawidg-doyen reviewed (changes requested):
I added some more comments in addition to the other ones provided on chat. Checkstyle linter should be run at the end too.

dawidg-doyen commented on google/detectors/credentials/generic_weak_credential_detector/src/test/java/com/google/tsunami/plugins/detectors/credentials/genericweakcredentialdetector/testers/rstudio/RStudioCredentialTesterTest.java:
```diff
@@ -0,0 +1,207 @@
+      TestCredential.create("root", Optional.of("pass"));
+  private static final TestCredential WRONG_CRED_1 =
+      TestCredential.create("wrong", Optional.of("pass"));
+  private static final ServiceContext.Builder rstudioServiceContext =
```
use capitals in the name if it's static

dawidg-doyen commented on google/detectors/credentials/generic_weak_credential_detector/src/test/java/com/google/tsunami/plugins/detectors/credentials/genericweakcredentialdetector/testers/rstudio/RStudioCredentialTesterTest.java:
```diff
@@ -0,0 +1,207 @@
+      }
+    }
+
+    
```
2 empty lines instead of 1 like above. Remove one.

leonardo-doyensec commented:
Updated everything as you requested. Regarding the `startsWith` i've changed it with a for loop on every `Set-Cookie` header and checks if one of them contains a strict `user-id` cookie
